core/node/state_keeper/src/keeper.rs

```diff
@@ -582,34 +594,52 @@ impl ZkSyncStateKeeper {
                     "L1 batch #{} should be sealed unconditionally as per sealing rules",
                     updates_manager.l1_batch.number
                 );
+
+                // Push the current block if it has not been done yet and this will effectively create a fictive l2 block
+                if let Some(next_l2_block_timestamp) = updates_manager.next_l2_block_timestamp_mut()
+                {
+                    self.io
+                        .update_next_l2_block_timestamp(next_l2_block_timestamp);
+                    Self::start_next_l2_block(updates_manager, batch_executor).await?;
+                }
+
                 return Ok(());
             }
 
-            if self.io.should_seal_l2_block(updates_manager) {
+            if !updates_manager.has_next_block_params()
+                && self.io.should_seal_l2_block(updates_manager)
+            {
```

Comment:
Do I understand it correctly that adding `!updates_manager.has_next_block_params()` is not strictly necessary here? I.e. `has_next_block_params()` can be true only if `updates_manager.l2_block.executed_transactions.is_empty()`. Just in case I'm not against adding it, just want to understand better

Reply:
if you don't have `!updates_manager.has_next_block_params()` then you might end up sealing the same block twice (=error) in case `wait_for_next_tx` timed out and the function loop over. 

(`start_next_l2_block` is called only when a transaction can be found in the mempool0

Reply:
I see, makes sense